### PR TITLE
Explore: Transform prometheus query to elasticsearch query

### PR DESCRIPTION
**Pull request by melchiormoulin**
**What this PR does / why we need it**:
Enable a way to transform prometheus/loki labels to elasticsearch query.
This make a link between metrics to logs.
Examples:
A prometheus query in Explore : `rate(my_metric{label1="value1",label2!="value2",label3=~"value.+",label4!~".*tothemoon"}[5m])`
Will be this elasticsearch query when switching to elasticsearch datasource:  `__name__:"my_metric" AND label1:"value1" AND NOT label2:"value2" AND label3:/value.+/ AND NOT label4:/.*tothemoon/`

### CODE REVIEW

**Comment by melchiormoulin** [comment]
Hello any thought about this @ivanahuckova @aocenas @davkal please ?  :)

**Comment by aocenas** [comment]
Hey @melchiormoulin thanks for contribution, this looks awesome. We are right before a big release right now so a bit short on time. Due to that it may take some time to review and merge, please bear with us.

Few question as I just quickly skimmed the code:
- Is this just one direction so queries from Prometheus to elastic and not the other way around?
- Based on the example I assume this transforms only the labels and metric portion of the Prometheus query and does not try to translate aggregation functions between those two, right?

**Comment by melchiormoulin** [comment]
Hello @aocenas 
Thank you for your answer :)

Is this just one direction so queries from Prometheus to elastic and not the other way around?
For now yes but it's possible to it. Great idea. May i do it here ?

Based on the example I assume this transforms only the labels and metric portion of the Prometheus query and does not try to translate aggregation functions between those two, right?
My idea was in case of investigations we want to grab the logs related to the metric and not display metrics based on logs in elasticsearch datasource.
Also all prometheus query can't be translated to elasticsearch query.
What do you think ?

**Comment by aocenas** [comment]
> Is this just one direction so queries from Prometheus to elastic and not the other way around?
For now yes but it's possible to it. Great idea. May i do it here ?

If you want to work on that I would prefer having it in separate PR if you don't mind, it can be branched off this PR and just mention that this needs to be dealt with first.

But in general I assume the most value is actually being able to go from prometheus metrics to elastic logs.

> My idea was in case of investigations we want to grab the logs related to the metric and not display metrics based on logs in elasticsearch datasource.
Also all prometheus query can't be translated to elasticsearch query.
What do you think ?

Yes I think this makes sense, as I said I also assume this has the most value. Also due to how the API is implemented right now each target datasource needs to understand syntax of the source datasource query. Because of that, from maintenance perspective it actually makes sense to limit the amount of syntax we transform until we are able to improve the general abstraction of this API.

**Comment by melchiormoulin** [comment]
> But in general I assume the most value is actually being able to go from prometheus metrics to elastic logs.
I agree. I let the PR like this, waiting for the review :)

**Comment by melchiormoulin** [comment]
any update on this please ? :)

**Comment by aocenas** [comment]
@melchiormoulin really sorry, this will have to wait for 7.1 release. Right now we have full hands with 7.0 which should be released next week and it is really hard to squeeze anything besides bugfixes in right now.

**Comment by melchiormoulin** [comment]
bye bot :D

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (aocenas) [inline public/app/plugins/datasource/elasticsearch/language_provider.ts:8]: Seems to work but I would like parsing to be based on the grammar we already have in https://github.com/grafana/grafana/blob/1448767c0822b790823b157a86073f8b3c171645/public/app/plugins/datasource/prometheus/promql.ts#L378

Basically you can do:
```
import Prism from 'prismjs';
import grammar from '../../prometheus/promql'

const tokens = Prism.tokenize(query, grammar);
```
And then iterate the tokens and check their types `label-key`, `label-value` and `operator`. Should make it easier to keep this in sync with any updates to promql and should be also less code. We use this approach around to do suggestions and query manipulation.

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (aocenas) [inline public/app/plugins/datasource/elasticsearch/language_provider.ts:86]: I do not think this and other functions here need to be async. Somewhere the interface requires a Promise to be returned but that can be done in the function that really requires that, here everything works sync anyway.

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (aocenas) [inline public/app/plugins/datasource/elasticsearch/language_provider.ts:93]: This could be a bit problematic because we do not actually change the mode in Explore based on this. So if you go from Prometheus to Elastic it will stay in Metrics mode but the queries will be done as if in Logs mode which may result in errors, though we have this issue without this PR. @kaydelaney does it make sense to do something about it or will this be moot anyway in 7.1 when we remove the mode switcher?

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (melchiormoulin) [reply public/app/plugins/datasource/elasticsearch/language_provider.ts:88]: done

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (melchiormoulin) [reply public/app/plugins/datasource/elasticsearch/language_provider.ts:95]: I don't get what do you mean, you can see the logs when you switch from prometheus to elasticsearch.

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (melchiormoulin) [reply public/app/plugins/datasource/elasticsearch/language_provider.ts]: done

**Comment by melchiormoulin** [comment]
Hello @aocenas :)
Is the tests failed because of my PR please ?

**Comment by tskarhed** [comment]
@melchiormoulin Not sure if the test fails because of it, but I think you need to merge master into this branch, as spellcheck is failing.

**Comment by ivanahuckova** [comment]
@melchiormoulin I have re-runed the test and it is passing now. 🙂 

Unfortunately, some of our E2E tests are sometimes flaky. We definitely want to fix that, but for now, re-running the test helped.

**Comment by melchiormoulin** [comment]
Hey @ivanahuckova thank you ! :D 
Do you know if there is any blocker to not merge this PR please ? :)
Have a nice day

File: public/app/plugins/datasource/elasticsearch/language_provider.ts
Comment (melchiormoulin) [reply public/app/plugins/datasource/elasticsearch/language_provider.ts:114]: Hey @aocenas 
Is it a blocker to not merge the PR please ?

**Comment by melchiormoulin** [comment]
ping ? :)

**Comment by ivanahuckova** [comment]
Hi melchiormoulin! I am really sorry for such a late reply. We are currently working on logs-metrics unification for elastic (and loki) that is going to be part of 7.1 release. Therefore we think it would make sense to wait with merging of/finalising this PR until then.

**Comment by melchiormoulin** [comment]
Hey @ivanahuckova, :) 
As this PR is ready to merge more than 1 month I'm not sure to understand why waiting.
Is there any PR to follow up the refacto?

**Comment by ivanahuckova** [comment]
Yes, here is the PR https://github.com/grafana/grafana/pull/25404.

**Comment by melchiormoulin** [comment]
Hello is it still relevant with https://github.com/grafana/grafana/pull/25890 ?

**Review by davkal — DISMISSED** [review]
this would be a great addition if we can get this to work. 

Questions:

- is the `metric` to `__name__:"metric"` conversion a good idea? I.e., is `__name__` a commonly used field with elastic users? If not, we can simply drop it.

Recommendations: 

- the label parsing should use [parseSelector](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/prometheus/language_utils.ts#L47) instead of relying on the Prism syntax.
- If you can get by with mocks for `TemplateSrv` and `TimeSrv` for the test class, then that would be preferable. The test class is already in good shape otherwise.
- Check if `isLogsQuery` is still needed after the logs/metrics unification PR.

**Comment by melchiormoulin** [comment]
Hello @davkal, thank you for your feedback.

> this would be a great addition if we can get this to work.
> 
> Questions:
> 
> * is the `metric` to `__name__:"metric"` conversion a good idea? I.e., is `__name__` a commonly used field with elastic users? If not, we can simply drop it.

The __name__ is the label key for the metric name in Prometheus,  this convention is also use in parseSelector function you mention.

I will do the other changes.

**Comment by melchiormoulin** [comment]
> Recommendations:
> 
> * the label parsing should use [parseSelector](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/prometheus/language_utils.ts#L47) instead of relying on the Prism syntax.
This method is only returning the labels keys and not label key + operator + value. Should i change the method signature to use it ?
> * If you can get by with mocks for `TemplateSrv` and `TimeSrv` for the test class, then that would be preferable. The test class is already in good shape otherwise.
Is it good now ?
> * Check if `isLogsQuery` is still needed after the logs/metrics unification PR.
After rebase seems it is still the case

**Comment by melchiormoulin** [comment]
Hello any news ?

**Comment by aocenas** [comment]
@melchiormoulin I am very sorry for the delays, there was a lot going on recently. I checked the code I think it looks good in general, wanted to do some minor changes and figure out why the build is failing but seems like I do not have permissions to push to your branch. Did you check "Allow edits from maintainers" when creating the PR?

**Comment by melchiormoulin** [comment]
Hello @aocenas the box allow edits by maintainers is style tick. I don't know why you can't edit 
Screenshot 2020-08-27 at 15 04 01

**Comment by davkal** [comment]
Could you help me clarify this a bit more: is the `metric` to `__name__:"metric"` conversion a good idea? I.e., is `__name__` a commonly used field with elastic users? If not, we can simply drop it.
You answer above simply states how `__name__` is a Prometheus label that names the metric (in effect syntactic sugar for `foo{} === {__name__: "foo"}`. But do metrics or logs stored in elastic have this label? I simply have not seen this, and I worry that if we allow `__name__` to be imported into the elastic query, we're adding a filter term that would return no results when sent as a query.

**Comment by melchiormoulin** [comment]
> Could you help me clarify this a bit more: is the `metric` to `__name__:"metric"` conversion a good idea? I.e., is `__name__` a commonly used field with elastic users? If not, we can simply drop it.
> You answer above simply states how `__name__` is a Prometheus label that names the metric (in effect syntactic sugar for `foo{} === {__name__: "foo"}`. But do metrics or logs stored in elastic have this label? I simply have not seen this, and I worry that if we allow `__name__` to be imported into the elastic query, we're adding a filter term that would return no results when sent as a query.

From my knowledge there is no convention for labels to link logs and metrics. The label __name__ can be indexed by elastic as the other labels. The idea of this PR was more to translate the all query, including the labels and the metric name and people can delete the labels that they don't have in their logs.  I don't have a strong opinion on this, i can remove it.

**Comment by aocenas** [comment]
Seems like the issue is with the naming, as you work on your master my local name has to be different and so you have to specify not to push to default branch name but specifically back to master.

**Comment by melchiormoulin** [comment]
So what do you think @aocenas @davkal , should i remove it the metric name in the elasticsearch query ?

**Comment by aocenas** [comment]
@davkal I am going to merge this. I think it would make sense to get some feedback and we can easily change the details later on. Personally I think it's easy to just delete stuff that does not match, better than starting from scratch and also I think the same problem can be with other labels as there is not mechanism to make them the same (as with loki and prometheus).

@melchiormoulin Thanks for your work and again sorry for being slow on this.

**Review by aocenas — APPROVED** [review]

**Pull request merged**
